### PR TITLE
[Backport] VectorAPI: 8223347: Integration of Vector API (Incubator) 

### PR DESCRIPTION
**Pull request by JinZhonghui**
[Backport] VectorAPI: 8223347: Integration of Vector API (Incubator) 

Summary: Backport VectorAPI 8223347: Integration of Vector API (Incubator)

Test Plan: ci jtreg

Reviewed-by: JoshuaZhuwj

Issue: https://github.com/alibaba/dragonwell11/issues/294

### CODE REVIEW

**Comment by JoshuaZhuwj** [comment]
I will begin corresponding tests and reviews. Please expect a delay in my feedback.
Thanks for your efforts.

**Comment by JoshuaZhuwj** [comment]
Some failures/crashes were found by the CI automation job.
The CI automation job of Dragonwell 11 comprises three components mainly: jdk tier1; hotspot tier1; langtools tier1.
I usually submit PR after no new failures are introduced in offline jtreg tests.
Offline jtreg tests are usually more strict than online CI verification.
What I run in my offline regression tests is comprised of: hotspot:hotspot_all_no_apps langtools:tier1 jdk:jdk_core.

File: src/hotspot/share/runtime/deoptimization.hpp [hunk 1]
```diff
@@ -159,7 +159,7 @@ class Deoptimization : AllStatic {
 JVMCI_ONLY(public:)
 
   // Support for restoring non-escaping objects
-  static bool realloc_objects(JavaThread* thread, frame* fr, GrowableArray<ScopeValue*>* objects, TRAPS);
```
Comment (JoshuaZhuwj) [inline src/hotspot/share/runtime/deoptimization.hpp:162]: It would be better keep this method and add 2nd method instead of direct replacement.

**Review by JoshuaZhuwj — COMMENTED** [review]
Thanks a lot for your efforts.
Please see my comments inline.

Besides the review comments, there exist building issues due to code modifications in platform-independent share codes. As a result, currently jdk cannot be built on other platforms. I think the building failure was introduced in previous commits before this large one in the dev branch.

I also found this large PR comprises many different commits of upstream, like below:
  8241042: x86_64: Improve Assembler generation
  8242492: Remove Matcher::vector_shift_count_ideal_reg()
  8239008: C2: Simplify Replicate support for sub-word types on x86
  8253191: C2: Masked byte comparisons with large masks produce wrong result on x86
  8230565: ZGC: Redesign C2 load barrier to expand on the MachNode level
Could you please not combine these different commits into this large PR?
Separate it into different commits just as did in upstream would be helpful for easy tracking in git logs in the future.

File: src/hotspot/share/opto/callGenerator.cpp [hunk 2]
```diff
@@ -289,16 +289,16 @@ CallGenerator* CallGenerator::for_virtual_call(ciMethod* m, int vtable_index) {
 // Allow inlining decisions to be delayed
 class LateInlineCallGenerator : public DirectCallGenerator {
  private:
-  // unique id for log compilation
-  jlong _unique_id;
+  jlong _unique_id;   // unique id for log compilation
+  bool _is_pure_call; // a hint that the call doesn't have important side effects to care about
```
[the remark this answers is not in the document]
Comment (JinZhonghui) [reply src/hotspot/share/opto/callGenerator.cpp:293]: Will remove these.

File: src/jdk.incubator.vector/share/classes/jdk/incubator/vector/gen-src.sh [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+}
+case $* in
+'')             CLASS_FILTER='*';;
+--generate*)    CLASS_FILTER=${2-'*'};;
+--clean)        MG=$(list_mech_gen); set -x; rm -f $MG; exit;;
+--revert)       MG=$(list_mech_gen); set -x; hg revert $MG; exit;;
+--list)         list_mech_gen; exit;;
+--help|*)       echo "Usage: $0 [--generate [file] | --clean | --revert | --list]"; exit 1;;
+esac
+
```
[the remark this answers is not in the document]
Comment (JinZhonghui) [reply src/jdk.incubator.vector/share/classes/jdk/incubator/vector/gen-src.sh:46]: Change -iinput_file -ooutput_file to < input_file > output_file in gen-src.sh

File: src/hotspot/cpu/x86/x86.ad [hunk 4]
```diff
@@ -5512,21 +7716,20 @@ instruct vfmaD_reg(vec a, vec b, vec c) %{
...
-    int vector_len = vector_length_encoding(this);
-    __ vfmad($c$$XMMRegister, $a$$XMMRegister, $b$$XMMRegister, $c$$XMMRegister, vector_len);
+    int vlen_enc = vector_length_encoding(this);
+    __ vfmad($c$$XMMRegister, $a$$XMMRegister, $b$$XMMRegister, $c$$XMMRegister, vlen_enc);
   %}
   ins_pipe( pipe_slow );
 %}
 
 instruct vfmaD_mem(vec a, memory b, vec c) %{
-  predicate(vector_length_in_bytes(n->in(1)) > 8);
```
[the remark this answers is not in the document]
Comment (JinZhonghui) [reply src/hotspot/cpu/x86/x86.ad:5522]: Looks like "predicate(UseAVX == 0);" is in the next code block (FMA). Please check it again.

File: src/hotspot/share/opto/matcher.cpp [hunk 5]
```diff
@@ -607,12 +645,25 @@ void Matcher::init_first_stack_mask() {
...
       idealreg2mhdebugmask[Op_RegD]->Remove(i);
-      idealreg2mhdebugmask[Op_RegP]->Remove(i);
+      idealreg2mhdebugmask[Op_VecZ]->Remove(i);
+
+      idealreg2mhdebugmask[Op_VecS]->Remove(i);
+      idealreg2mhdebugmask[Op_VecD]->Remove(i);
+      idealreg2mhdebugmask[Op_VecX]->Remove(i);
+      idealreg2mhdebugmask[Op_VecY]->Remove(i);
+      idealreg2mhdebugmask[Op_RegD]->Remove(i);
+
```
[the remark this answers is not in the document]
Comment (JinZhonghui) [reply src/hotspot/share/opto/matcher.cpp:666]: Fixed.

File: src/hotspot/share/opto/matcher.cpp [hunk 6]
```diff
@@ -427,7 +427,7 @@ static RegMask *init_input_masks( uint size, RegMask &ret_adr, RegMask &fp ) {
 void Matcher::init_first_stack_mask() {
 
   // Allocate storage for spill masks as masks for the appropriate load type.
-  RegMask *rms = (RegMask*)C->comp_arena()->Amalloc_D(sizeof(RegMask) * (3*6+5));
+  RegMask *rms = (RegMask*)C->comp_arena()->Amalloc_D(sizeof(RegMask) * (3*6+5*3));
```
[the remark this answers is not in the document]
Comment (JinZhonghui) [reply src/hotspot/share/opto/matcher.cpp:430]: Changed as your comments.

**Pull request closed without merging**